XenAdmin/Dialogs/NewTagDialog.cs

```diff
@@ -162,257 +83,137 @@ private void textBox1_KeyPress(object sender, KeyPressEventArgs e)
             }
         }
 
-        private void SortList()
+        private List<TagsDataGridViewRow> ExtractList()
         {
-            List<TagsListViewItem> listChecked = new List<TagsListViewItem>();
-            List<TagsListViewItem> listIndeterminate = new List<TagsListViewItem>();
-            List<TagsListViewItem> listNonChecked = new List<TagsListViewItem>();
-
-            foreach (TagsListViewItem item in tagsListView.Items)
+            var rows = new List<TagsDataGridViewRow>();
+            foreach (TagsDataGridViewRow item in tagsDataGrid.Rows)
             {
-                if (item.Checked == CheckState.Checked)
-                {
-                    listChecked.Add(item);
-                }
-                else if (item.Checked == CheckState.Unchecked)
-                {
-                    listNonChecked.Add(item);
-                }
-                else
-                {
-                    listIndeterminate.Add(item);
-                }
-            }
-            listNonChecked.Sort(Comparison());
-            listChecked.Sort(Comparison());
-            listIndeterminate.Sort(Comparison());
-            tagsListView.Items.Clear();
-            foreach (TagsListViewItem item in listChecked)
-            {
-                tagsListView.Items.Add(item);
-            }
-            foreach (TagsListViewItem item in listIndeterminate)
-            {
-                tagsListView.Items.Add(item);
-            }
-            foreach (TagsListViewItem item in listNonChecked)
-            {
-                tagsListView.Items.Add(item);
+                rows.Add(item);
             }
+            return rows;
         }
 
-        private Comparison<TagsListViewItem> Comparison()
+        private void DisplayList(List<TagsDataGridViewRow> rows)
         {
-            return delegate(TagsListViewItem x, TagsListViewItem y)
-                       {
-                           return x.Text.CompareTo(y.Text);
+            rows.Sort();
 
-                       };
+            tagsDataGrid.Rows.Clear();

```

Comment:
Normally when we clear/re-add rows in a datagridview we do it in a try-optinally catch-finally block as follows to prevent display problems:
```
try
{
    tagsDataGrid.SuspendLayout();
    tagsDataGrid.Rows.Clear();

    foreach (var item in rows)
        tagsDataGrid.Rows.Add(item);
}
finally
{
    tagsDataGrid.ResumeLayout();
}
```